modulemd/modulemd-module.c

```diff
@@ -751,6 +767,67 @@ modulemd_module_get_translation (ModulemdModule *self, const gchar *stream)
 }
 
 
+GPtrArray *
+modulemd_module_get_eol_keys (ModulemdModule *self)
+{
+  return modulemd_ordered_str_keys (self->eol, modulemd_strcmp_sort);
+}
+
+
+ModulemdEol *
+modulemd_module_get_eol (ModulemdModule *self, const gchar *key)
+{
+  return g_hash_table_lookup (self->eol, key);
+}
+
+
+void
+modulemd_module_add_eol (ModulemdModule *self, ModulemdEol *eol)
+{
+  gsize i;
+  ModulemdModuleStream *stream = NULL;
+  ModulemdEol *neweol = NULL;
+
+  g_return_if_fail (g_str_equal (modulemd_eol_get_module_name (eol),
+                                 modulemd_module_get_module_name (self)));
+
+  neweol = modulemd_eol_copy (eol);
+
+  const gchar *stream_str = modulemd_eol_get_module_stream (neweol);
+  const gchar *context_str = modulemd_eol_get_module_context (neweol);
+
+  g_hash_table_replace (
+    self->eol, g_strjoin (":", stream_str, context_str, NULL), neweol);
+
+  for (i = 0; i < self->streams->len; i++)
+    {
+      stream = (ModulemdModuleStream *)g_ptr_array_index (self->streams, i);
+
+      if (!g_str_equal (stream_str,
+                        modulemd_module_stream_get_stream_name (stream)))
+        {
+          continue;
+        }
+
+      /* If no context specified set neweol for all matching stream */
+
+      if (context_str)
+        {
+          if (!g_str_equal (context_str,
+                            modulemd_module_stream_get_context (stream)))
+            {
+              continue;
+            }
+          modulemd_module_stream_associate_eol (stream, neweol);
+        }
+      else
+        {
+          modulemd_module_stream_associate_eol (stream, neweol);
+        }
+    }
```

Comment:
I think you need to define the behavior for a mixture of documents that do or do not have context specified.

For example, as written right now, the result of having the following two `modulemd-eol` documents will behave differently depending on the order they appear in the modulemd YAML stream:

```yaml
document: modulemd-eol
version: 1
data:
  name: nodejs
  stream: 10
  context: deadbeef
  obsoleted_by:
    module: nodejs
    stream: 12
```
vs.
```yaml
document: modulemd-eol
version: 1
data:
  name: nodejs
  stream: 10
  obsoleted_by:
    module: nodejs
    stream: 14
```

Read in the order above, the result would be that `nodejs:10` was obsoleted by `nodejs:14` (because the no-context document would overwrite the context-including one that was read earlier).

Reverse the order and the result would be that `nodejs:10:deadbeef` is obsoleted by `nodejs:12` while any other context of `nodejs:10` will be obsoleted by `nodejs:14`.

Please figure out how to do this without ordering differences.

Reply:
Thanks for the batch of feedback, I have pushed an update.

I have also added a modified date check, so that the associated eol is the newest one. If there are identical dates it prefers the more specific (with context) eol.
This way the ordering problem still happens if there are conflicting eols with the same modified date, module, stream and context values, so I have added `g_info` message for this case.

I will also add this logic to merging indexes.

Reply:
Added the logic to merging indexes.